MochiKit/async/Deferred.js

```diff
@@ -0,0 +1,308 @@
+    return this.then(null, errfunc);
+};
+
+Deferred.prototype.finally = function(func) {
+    if (typeof func === 'function') {
+        this.addCallback(func, SETTLED);
+    }
+
+    return this;
+};
+
+Deferred.prototype.flip = function() {
+    let newChain = [];
+    for (let [callback, state] of this.chain) {
+        if (state === REJECTED) {
+            newChain.push([callback, RESOLVED]);
+        } else if (state === RESOLVED) {
+            newChain.push([callback, REJECTED]);
+        } else {
+            //We have a SETTLED state here,
+            //just keep it as that.
+            newChain.push([callback, state]);
+        }
+    }
+
+    this.chain = newChain;
+    return this;
+};
+
+Deferred.prototype.addCallback = function(callback, state) {
+    if (state !== RESOLVED && state !== REJECTED && state !== SETTLED) {
+        throw new Error('The provided state is not a valid state.');
+    }
+
+    if (this.state === state) {
+        //The state matches.
+        //Fire the callback.
+        this.fireCallback(callback);
+    } else {
+        this.chain.push([callback, state]);
+    }
+
+    return this;
+};
+
+Deferred.prototype.fireCallbacks = function() {
+    for (let [callback, state] of this.chain) {
+        if (this.correctState(state)) {
+            this.fireCallback(callback);
+        }
+    }
+
+    return this;
+};
+
+Deferred.prototype.correctState = function(state) {
+    return (
+        (state === REJECTED && this.isRejected()) ||
+        (state === RESOLVED && this.isResolved()) ||
+        (state === SETTLED && this.isSettled())
+    );
+};
+
+Deferred.prototype.fireCallback = function(callback) {
+    this.value = callback(this.value);
+    return this;
+};
+
+Deferred.prototype.resolve = function(value) {
+    this.value = value;
+    this.state = RESOLVED;
+    this.fireCallbacks();
+};
+
+Deferred.prototype.rejectAfter = function(timeout, value) {
+    return this.actionAfter('reject', timeout, value);
+};
+
+Deferred.prototype.resolveAfter = function(timeout, value) {
+    return this.actionAfter('resolve', timeout, value);
+};
+
+Deferred.prototype.actionAfter = function(action, timeout, value) {
+    let self = this,
+        method = action === REJECTED ? 'reject' : 'resolve';
+
+    setTimeout(function(timestamp) {
+        self[method](value === undefined ? timestamp : value);
+    }, timeout);
+};
+
+Deferred.prototype.catchSilent = function(errfunc) {
+    return this.catch(function() {});
+};
+
+Deferred.prototype.tapCatch = function(errfunc) {
+    if (this.state === REJECTED) {
+        errfunc(this.value);
+    } else {
+        this.__error_handlers__.push(errfunc);
+    }
+
+    return this;
+};
+
+Deferred.prototype.tap = function(func) {
+    return this.then(function(value) {
+        //Prevent the value from being changed
+        //by the callback, as it is a lazy function
+        //meaning it performs operations without
+        //returning a value.
+        func(value);
```

Comment:
A slightly different semantic for 'tap' is that if 'func' returns a Promise the chain waits for it. Bluebird does that for example.

Reply:
Interesting. This was heavily inspired by Bluebird,